src/core/metadata/qgslayermetadata.cpp

```diff
@@ -273,6 +278,460 @@ void QgsLayerMetadata::readFromLayer( const QgsMapLayer *layer )
+  title.appendChild( titleText );
+  metadataElement.appendChild( title );
+
+  // abstract
+  QDomElement abstract = document.createElement( QStringLiteral( "abstract" ) );
+  QDomText abstractText = document.createTextNode( mAbstract );
+  abstract.appendChild( abstractText );
+  metadataElement.appendChild( abstract );
+
+  // keywords
+  QMapIterator<QString, QStringList> i( mKeywords );
+  while ( i.hasNext() )
+  {
+    i.next();
+    QDomElement keywordsElement = document.createElement( QStringLiteral( "keywords" ) );
+    keywordsElement.setAttribute( QStringLiteral( "vocabulary" ), i.key() );
+    const QStringList values = i.value();
+    for ( const QString &kw : values )
+    {
+      QDomElement keyword = document.createElement( QStringLiteral( "keyword" ) );
+      QDomText keywordText = document.createTextNode( kw );
+      keyword.appendChild( keywordText );
+      keywordsElement.appendChild( keyword );
+    }
+    metadataElement.appendChild( keywordsElement );
+  }
+
+  // fees
+  QDomElement fees = document.createElement( QStringLiteral( "fees" ) );
+  QDomText feesText = document.createTextNode( mFees );
+  fees.appendChild( feesText );
+  metadataElement.appendChild( fees );
+
+  // constraints
+  for ( const QgsLayerMetadata::Constraint &constraint : mConstraints )
+  {
+    QDomElement constraintElement = document.createElement( QStringLiteral( "constraints" ) );
+    constraintElement.setAttribute( QStringLiteral( "type" ), constraint.type );
+    QDomText constraintText = document.createTextNode( constraint.constraint );
+    constraintElement.appendChild( constraintText );
+    metadataElement.appendChild( constraintElement );
+  }
+
+  // rights
+  for ( const QString &right : mRights )
+  {
+    QDomElement rightElement = document.createElement( QStringLiteral( "rights" ) );
+    QDomText rightText = document.createTextNode( right );
+    rightElement.appendChild( rightText );
+    metadataElement.appendChild( rightElement );
+  }
+
+  // license
+  for ( const QString &license : mLicenses )
+  {
+    QDomElement licenseElement = document.createElement( QStringLiteral( "license" ) );
+    QDomText licenseText = document.createTextNode( license );
+    licenseElement.appendChild( licenseText );
+    metadataElement.appendChild( licenseElement );
+  }
+
+  // encoding
+  QDomElement encoding = document.createElement( QStringLiteral( "encoding" ) );
+  QDomText encodingText = document.createTextNode( mEncoding );
+  encoding.appendChild( encodingText );
+  metadataElement.appendChild( encoding );
+
+  // crs
+  QDomElement crsElement = document.createElement( QStringLiteral( "crs" ) );
+  mCrs.writeXml( crsElement, document );
+  metadataElement.appendChild( crsElement );
+
+  // extent
+  QDomElement extentElement = document.createElement( QStringLiteral( "extent" ) );
+
+  // spatial extents
+  const QList< QgsLayerMetadata::SpatialExtent > &sExtents = extent().spatialExtents();
```

Comment:
This looks dangerous - you're storing a reference to a temporary (since spatialExtents() returns a new object). If you remove the reference here it will be safe and fast (due to the implicit sharing of the container, there's very minimal benefit to this being a reference anyway) . Same applies for temporalExtents() below.